### DISCUSSION TITLE
Temple apa7

### DISCUSSION THREAD

**Discussion by alsonicr** [quarto-dev/quarto-cli#4019] [Q&A]
Hello,

i'm trying to create a  pdf latex template for the document class _apa7_.

However the handling of the authors is not typical from other latex template (see https://ctan.math.illinois.edu/macros/latex/contrib/apa7/apa7.pdf for details)
```
\author{⟨author name⟩}
\twoauthors{First Author(s)}{Second Author(s)}
\threeauthors{John and Jim}{Mary and Sue}{Nick}
\fourauthors{Helen}{Dick}{Tracy and Larry}{James Bond}
```

Therefore quarto yaml metadata _author_ is not working 

My first ideas was to use the length of author meta data and could have look like that : 
```
$if length(author)==3$  
\threeauthors{John and Jim}{Mary and Sue}{Nick}
$else$ $if length(author)==2$
\twoauthors{First Author(s)}{Second Author(s)}
....
```
but could find a way to access author length or to access a specific rank in author.

```
My  second  ideas was to create to create a metadata author1 to author6 and add :
```
`$if (author2)$`
\twoauthors{John  Jim}{Mary  Sue}
$endif$
```
and could not make it work.

does someone have any ideas ?

**Comment by cderv** [comment]
Regarding Journal Template, you'll find a list of existing example in https://quarto.org/docs/extensions/listing-journals.html

There will be example there of how authors are handled in the . It is is not just a simple `authors`  value, Quarto expose a normalized version of this metadata. See more in https://quarto.org/docs/journals/authors.html
If you're not leveraging the `by-authors` meta, maybe it can help. 

`length` is a good idea but I don't think Pandoc's templating system as conditional for this testing of equality. Conditional is only for testing variable existence and truthiness. 

I'll think more if we have some ways to express this iteration when a template require different command according to the number of authors. Not something we have encountered yet I think. Did you @dragonstyle ?

**Reply by dragonstyle** [reply]
I haven't encountered this exact issue, but I do recall seeing this type of approach. I think the templating language itself will be too limiting to solve this well, but I think it could be solver with a LUA filter which pre-processed the authors into the correct form, which can then be blinding output in the template. Something like:

**filter.lua**
```lua
Meta = function(meta)
  local byAuthor = meta['by-author']
  local apaAuthorPfx = '\\author'
  if byAuthor ~= nil then

    if #byAuthor == 2 then
      apaAuthorPfx = '\\twoauthors'
    elseif #byAuthor == 3 then
      apaAuthorPfx = "\\threeauthors"
    end
    -- TODO: 4, 5, 6

    meta['apa-author-prefix'] = apaAuthorPfx
  end
end
```

and in the template:

**title.tex**
```tex
$apa-author-prefix$$for(by-author)${$by-author.name.literal$}$endfor$
```

**Reply by alsonicr** [reply]
Thank for the help. 

I'm not familiar whit lua and i must have miss something. 

I had the title.tex in the template-partials :

```
title: Apa7
author: First Last
version: 1.0.0
quarto-required: ">=1.2.0"
contributes:
  formats:
    common:
      toc: true
      filters:
        - apa7.lua
    pdf:
      documentclass: apa7
      include-in-header: header.tex
      template-partials:
        - partials/title.tex
      
```
and the suggested lua in the apa7.lua for testing

the names of the authors are provided in the latex file when running quarto but with not command attach see : 

```
\hypersetup{
  pdftitle={Apa7 Template},
  pdfauthor={Sarah Malloc; Eliza Dealloc},
  pdfkeywords={template, demo},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

**{Sarah Malloc}{Eliza Dealloc}**
\begin{document}
```

I don't get why the $apa-author-prefix$ is not provided

**Reply by cderv** [reply]
What is in your partials title.tex ? Or rather can you build a small github repo with your testing so that we work on the same thing ? It would be easier to debug

**Reply by alsonicr** [reply]
`$apa-author-prefix$$for(by-author)${$by-author.name.literal$}$endfor$ `has mention and suggested by @dragonstyle  above

Yes i will make git

**Reply by dragonstyle** [reply]
There is a bug in the LUA code that I provided! It should read:


```lua
Meta = function(meta)
  local byAuthor = meta['by-author']
  local apaAuthorPfx = '\\author'
  if byAuthor ~= nil then

    if #byAuthor == 2 then
      apaAuthorPfx = '\\twoauthors'
    elseif #byAuthor == 3 then
      apaAuthorPfx = "\\threeauthors"
    end
    -- TODO: 4, 5, 6

    meta['apa-author-prefix'] = apaAuthorPfx
    return meta --< This was missing! without it, Pandoc will just use the unmodified `meta` value
  end
end
```

**Reply by alsonicr** [reply]
mhmm it seems the render get this 

`\textbackslash twoauthors{Sarah Malloc}{Eliza Dealloc}`
the `\\` in lua has been translated by `\textbackslash` and not `\`


and i did create a git repo here [https://github.com/alsonicr/quarto-apa7](https://github.com/alsonicr/quarto-apa7)

**Reply by cderv** [reply]
it seems Meta value will be parsed as Markdown. Even from Lua. I wonder if you could do this then 

````lua
Meta = function(meta)
  local byAuthor = meta['by-author']
  local apaAuthorPfx = '`\\author`{=latex}'
  if byAuthor ~= nil then

    if #byAuthor == 2 then
      apaAuthorPfx = '`\\twoauthors`{=latex}'
    elseif #byAuthor == 3 then
      apaAuthorPfx = "`\\threeauthors`{=latex}"
    end
    -- TODO: 4, 5, 6

    meta['apa-author-prefix'] = apaAuthorPfx
    return meta --< This was missing! without it, Pandoc will just use the unmodified `meta` value
  end
end
````

or maybe it is just `\\` escaping is not needed and single could be used `\`

To try in your demo repo... This is really interesting !

**Reply by alsonicr** [reply]
If we used only `\` we get `woauthors{Sarah Malloc}{Eliza Dealloc}`

if we use the `{=latex} wee get 

`\textbackslash twoauthors`\{=latex\}{Sarah Malloc}{Eliza Dealloc}

**Reply by cderv** [reply]
ok thank you. Let me have a look with your example

**Reply by cderv** [reply]
So we need to really do as we do in non Meta object: used the AST representation for the raw latex part. 

````lua
Meta = function(meta)
  local byAuthor = meta['by-author']
  local apaAuthorPfx = pandoc.RawInline('latex', '\\author')
  if byAuthor ~= nil then

    if #byAuthor == 2 then
      apaAuthorPfx = pandoc.RawInline('latex', '\\twoauthors')
    elseif #byAuthor == 3 then
      apaAuthorPfx = pandoc.RawInline('latex', '\\threeauthors')
    end
    -- TODO: 4, 5, 6

    meta['apa-author-prefix'] = apaAuthorPfx
    return meta --< This was missing! without it, Pandoc will just use the unmodified `meta` value
  end
end
````

However, just a heads up that this will throw an error with latest **apa7** from CTAN as the commands are now deprecated. 

````powershell
compilation failed- error
Class apa7 Error: Deprecated: see documentation for \authorsnames.

See the apa7 class documentation for explanation.
````

> However the handling of the authors is not typical from other latex template (see [ctan.math.illinois.edu/macros/latex/contrib/apa7/apa7.pdf](https://ctan.math.illinois.edu/macros/latex/contrib/apa7/apa7.pdf) for details)

From the document you linked here, it says it is deprecated. 

![image](https://user-images.githubusercontent.com/6791940/213750556-f03bff60-c94d-41af-ad82-fd4ae33bc3f5.png)

Command to use is `\authorsnames{}` now 

See https://mirrors.nxthost.com/ctan/macros/latex/contrib/apa7/apa7.pdf

**Reply by alsonicr** [reply]
I change the authors and add a title and short title in the repo and it work
There is a small problem with the abstract, i will try to solve it and keep you update.

thanks for the help !

**Reply by alsonicr** [reply]
@dragonstyle I recently tried to add table in the two-column journal classoption.

I have seen you workaround [here ](https://github.com/quarto-dev/quarto-cli/discussions/2786#discussioncomment-3853259).

The solution seems to "work" (produce a pdf) sometimes and sometimes not, at random. I couldn't replicate when it does or doesn't work, because it happens when I don't change any file at all (e.g. I push the commit like it "works" and it doesn't).

Have I forgotten something for the `_two-column-longtable.tex` solution? For example, I did not find it referenced in the partial template from Elsevier.

**Reply by dragonstyle** [reply]
I'm not sure - in the Elsevier case it may be that there aren't two column formats or that the Elsevier class can accept long tables. Without seeing the exact tex generated, errors, and so on it is pretty difficult to diagnose. If you can try to narrow it down to something repeatable I am happy to have a look...

**Reply by alsonicr** [reply]
here is the error 

```
compilation failed- error
Package longtable Error: longtable not in 1-column mode.

See the longtable package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.249 \begin{longtable}
                       []{@{}llrc@{}} 

see template.log for more information.
```
template.log return the same.

Normally it should be able to replicate the error from the repos :[ https://github.com/alsonicr/quarto-apa7]( https://github.com/alsonicr/quarto-apa7)

Normally ...

**Reply by dragonstyle** [reply]
Yes I could reproduce it - you need to actually include / use that `.tex` file. For example, I moved it out of the templates directory and up a level and then updated the `_extension.yml` file to read:

```markdown
title: Apa7
author: First Last
version: 1.0.0
quarto-required: ">=1.2.0"
contributes:
  formats:
    common:
      toc: true
      filters:
        - apa7.lua
    pdf:
      documentclass: apa7
      include-in-header: _two-column-longtable.tex
      template-partials:
        - "partials/before-body.tex"
        - "partials/title.tex"
        
      
    # html:
    #   css: styles.css
```

This is the basic idea - include this tex file in the header when you're in 2 column mode. There are other fancier ways to handle this (for example from within LUA detecting whether the format will use 2 columns or not) but I'm not sure whether that is necessary here.

**Reply by alsonicr** [reply]
It seems to work

However, for some reason I can't seem to get the `_two-column-longtable.tex` code to interact with the Latex production. If I don't put a large chunk of text before the table, the text before the table is not included in the PDF. But if I remove the table of contents (`toc: false`) it works fine. Any ideas why? (sometimes latex compilation seems to loop infitly since I set max run to 20)


When `classoption: man` for the manuscript pdf other latex code is not included in the pdf when longtable start. My best guess is that `_two-column-longtable.tex` should only be included in the header when the `jou` option is used with some lua as you mention.

I was thinking of something like that:
```
Meta = function(meta)
  -- classoption function
  local test = meta['classoption']
  if test[1]=="jou" then 
    print("## APA journal format ##")  
    longtable = pandoc.RawInline('latex', '\include{"_two-column-longtable.tex"}')
    meta['longtable'] = longtable
    return meta
  else  
    print("### APA manuscript format###")  
  end
end
```
But I am not sure how to test the classoption `man` or `jou` 

repo is update
